File: clients/go/decode.go
// Copyright 2026 Bany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package bigdb

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strconv"
)

// RawResponse in, typed answer or typed error out. No I/O here, which is what makes the whole
// of it testable against bytes copied out of the server's own tests.

// decodeError turns a non-2xx response into a *ServerError.
//
// The envelope is {"error": <sentence>, "code": <code>} - big_http::json::error. An older
// document elsewhere in this repository calls the sentence "message"; that is wrong, and this
// reads "error" first and falls back to "message" rather than requiring either, because a body
// that is neither still has a status worth reporting.
func decodeError(r *RawResponse) *ServerError {
	var body struct {
		Error   string `json:"error"`
		Message string `json:"message"`
		Code    string `json:"code"`
	}
	_ = json.Unmarshal(r.Body, &body)

	msg := body.Error
	if msg == "" {
		msg = body.Message
	}
	if msg == "" {
		// A 5xx body is redacted, so there may be nothing here at all. Say so plainly rather
		// than reporting an empty sentence.
		msg = "the server gave no detail"
		if len(r.Body) > 0 && len(r.Body) < 200 {
			msg = string(bytes.TrimSpace(r.Body))
		}
	}
	e := &ServerError{
		Status:    r.Status,
		Code:      body.Code,
		Message:   msg,
		RequestID: r.RequestID(),
	}
	if d, ok := r.RetryAfter(); ok {
		e.RetryAfter = d
	}
	return e
}

func unmarshalObject(b []byte) (map[string]json.RawMessage, error) {
	var m map[string]json.RawMessage
	if err := json.Unmarshal(b, &m); err != nil {
		return nil, &ProtocolError{What: "the answer is not a JSON object: " + err.Error()}
	}
	return m, nil
}

func asUint64(raw json.RawMessage, what string) (uint64, error) {
	n, err := strconv.ParseUint(string(bytes.TrimSpace(raw)), 10, 64)
	if err != nil {
		return 0, &ProtocolError{What: fmt.Sprintf("%s is not a u64: %s", what, raw)}
	}
	return n, nil
}

func asInt64(raw json.RawMessage, what string) (int64, error) {
	n, err := strconv.ParseInt(string(bytes.TrimSpace(raw)), 10, 64)
	if err != nil {
		return 0, &ProtocolError{What: fmt.Sprintf("%s is not an i64: %s", what, raw)}
	}
	return n, nil
}

func isNull(raw json.RawMessage) bool {
	return string(bytes.TrimSpace(raw)) == "null"
}

// hasPoint reports whether a JSON number was written as a real rather than as an integer.
//
// big_http::json::real always writes a point ("a column that is sometimes 3 and sometimes 3.5
// is a column a client has to sniff"), so this is a reliable read of which kind of sum came
// back rather than a guess.
func hasPoint(raw json.RawMessage) bool {
	return bytes.ContainsAny(raw, ".eE")
}

// decodeAnswer reads one PQL answer.
//
// The keys are checked in a fixed order because the wire form has no discriminant. A body that
// matches none of them is a *ProtocolError: a shape this build has never heard of should be
// said out loud, not returned as a map wearing a type.
func decodeAnswer(body []byte) (Answer, error) {
	m, err := unmarshalObject(body)
	if err != nil {
		return nil, err
	}
	raw := json.RawMessage(body)

	if v, ok := m["count"]; ok {
		n, err := asUint64(v, "count")
		if err != nil {
			return nil, err
		}
		return &CountAnswer{Count: n, Raw: raw}, nil
	}

	if v, ok := m["sum"]; ok {
		a := &SumAnswer{Raw: raw}
		if hasPoint(v) {
			a.IsReal = true
			if err := json.Unmarshal(v, &a.Real); err != nil {
				return nil, &ProtocolError{What: "sum is not a number: " + string(v)}
			}
			return a, nil
		}
		if a.Value, err = asInt64(v, "sum"); err != nil {
			return nil, err
		}
		return a, nil
	}

	if v, ok := m["value"]; ok {
		a := &ValueAnswer{Raw: raw}
		if isNull(v) {
			return a, nil
		}
		a.Present = true
		if hasPoint(v) {
			a.IsReal = true
			if err := json.Unmarshal(v, &a.Real); err != nil {
				return nil, &ProtocolError{What: "value is not a number: " + string(v)}
			}
			return a, nil
		}
		if a.Value, err = asInt64(v, "value"); err != nil {
			return nil, err
		}
		return a, nil
	}

	if _, ok := m["records"]; ok {
		ids, next, err := decodeRecordList(m)
		if err != nil {
			return nil, err
		}
		return &RecordsAnswer{Records: ids, Next: next, Raw: raw}, nil
	}

	if v, ok := m["tuples"]; ok {
		var items []struct {
			Keys  []json.RawMessage `json:"keys"`
			Value json.RawMessage   `json:"value"`
		}
		if err := json.Unmarshal(v, &items); err != nil {
			return nil, &ProtocolError{What: "tuples is not a list: " + err.Error()}
		}
		out := make([]TupleItem, 0, len(items))
		for _, it := range items {
			keys := make([]GroupKey, 0, len(it.Keys))
			for _, k := range it.Keys {
				gk, err := decodeGroupKey(k)
				if err != nil {
					return nil, err
				}
				keys = append(keys, gk)
			}
			inner, err := decodeAnswer(it.Value)
			if err != nil {
				return nil, err
			}
			out = append(out, TupleItem{Keys: keys, Value: inner})
		}
		return &TuplesAnswer{Tuples: out, Raw: raw}, nil
	}

	if v, ok := m["groups"]; ok {
		var items []json.RawMessage
		if err := json.Unmarshal(v, &items); err != nil {
			return nil, &ProtocolError{What: "groups is not a list: " + err.Error()}
		}
		out := make([]GroupItem, 0, len(items))
		for _, it := range items {
			gk, err := decodeGroupKey(it)
			if err != nil {
				return nil, err
			}
			var wrap struct {
				Value json.RawMessage `json:"value"`
			}
			if err := json.Unmarshal(it, &wrap); err != nil {
				return nil, &ProtocolError{What: "a group is not an object: " + err.Error()}
			}
			inner, err := decodeAnswer(wrap.Value)
			if err != nil {
				return nil, err
			}
			out = append(out, GroupItem{GroupKey: gk, Value: inner})
		}
		return &GroupsAnswer{Groups: out, Raw: raw}, nil
	}

	if v, ok := m["rows"]; ok {
		var items []struct {
			Record json.RawMessage   `json:"record"`
			Values []json.RawMessage `json:"values"`
		}
		if err := json.Unmarshal(v, &items); err != nil {
			return nil, &ProtocolError{What: "rows is not a list: " + err.Error()}
		}
		out := make([]ProjectionRow, 0, len(items))
		for _, it := range items {
			id, err := asUint64(it.Record, "record")
			if err != nil {
				return nil, err
			}
			out = append(out, ProjectionRow{Record: id, Values: it.Values})
		}
		return &RowsAnswer{Rows: out, Raw: raw}, nil
	}

	return nil, &ProtocolError{What: "this answer is a shape this client does not know: " +
		truncate(string(body), 200)}
}

func decodeGroupKey(b []byte) (GroupKey, error) {
	var k struct {
		Key json.RawMessage `json:"key"`
		Row json.RawMessage `json:"row"`
	}
	if err := json.Unmarshal(b, &k); err != nil {
		return GroupKey{}, &ProtocolError{What: "a group key is not an object: " + err.Error()}
	}
	out := GroupKey{}
	if len(k.Key) > 0 && !isNull(k.Key) {
		if err := json.Unmarshal(k.Key, &out.Key); err != nil {
			return GroupKey{}, &ProtocolError{What: "a group key is not a string: " + string(k.Key)}
		}
		out.HasKey = true
	}
	row, err := asInt64(k.Row, "row")
	if err != nil {
		return GroupKey{}, err
	}
	out.Row = row
	return out, nil
}

// decodeRecordList reads the {"records": [...], "next": id|null} pair, which two routes share.
func decodeRecordList(m map[string]json.RawMessage) ([]uint64, *uint64, error) {
	var rawIDs []json.RawMessage
	if err := json.Unmarshal(m["records"], &rawIDs); err != nil {
		return nil, nil, &ProtocolError{What: "records is not a list: " + err.Error()}
	}
	// Decoded one at a time through ParseUint rather than into []uint64 wholesale, so that a
	// value that is not an integer names itself instead of coming back as a zero.
	ids := make([]uint64, 0, len(rawIDs))
	for _, r := range rawIDs {
		id, err := asUint64(r, "a record id")
		if err != nil {
			return nil, nil, err
		}
		ids = append(ids, id)
	}

	var next *uint64
	if v, ok := m["next"]; ok && !isNull(v) {
		n, err := asUint64(v, "next")
		if err != nil {
			return nil, nil, err
		}
		next = &n
	}
	return ids, next, nil
}

// decodeRecordPage reads GET /table/{t}/records.
func decodeRecordPage(body []byte) (*RecordPage, error) {
	m, err := unmarshalObject(body)
	if err != nil {
		return nil, err
	}
	if _, ok := m["records"]; !ok {
		return nil, &ProtocolError{What: "a record page with no records key"}
	}
	ids, next, err := decodeRecordList(m)
	if err != nil {
		return nil, err
	}
	return &RecordPage{Records: ids, Next: next, Raw: body}, nil
}

// decodeSQL reads POST /sql, deciding on the response's Content-Type and nothing else.
func decodeSQL(r *RawResponse) (*SQLResult, error) {
	if ct := r.ContentType(); ct != "application/json" {
		return &SQLResult{Text: r.Body, ContentType: ct}, nil
	}
	var set struct {
		Columns []string            `json:"columns"`
		Rows    [][]json.RawMessage `json:"rows"`
	}
	if err := json.Unmarshal(r.Body, &set); err != nil {
		return nil, &ProtocolError{What: "the result set is not readable: " + err.Error()}
	}
	if set.Columns == nil {
		set.Columns = []string{}
	}
	return &SQLResult{Columns: set.Columns, Rows: set.Rows, Raw: r.Body}, nil
}

// decodeWrite reads the {"<name>": n} or {"<name>": n, "missed": [...]} that an import or a
// delete answers with. See big_http::json::wrote.
func decodeWrite(body []byte, name string) (*WriteResult, error) {
	m, err := unmarshalObject(body)
	if err != nil {
		return nil, err
	}
	v, ok := m[name]
	if !ok {
		return nil, &ProtocolError{What: "a write answer with no " + name + " key"}
	}
	n, err := asUint64(v, name)
	if err != nil {
		return nil, err
	}
	out := &WriteResult{Count: n, Raw: body}
	if raw, ok := m["missed"]; ok {
		if err := json.Unmarshal(raw, &out.Missed); err != nil {
			return nil, &ProtocolError{What: "missed is not a list of strings: " + err.Error()}
		}
	}
	return out, nil
}

// decodeSchema reads GET /schema.
func decodeSchema(body []byte) (*Schema, error) {
	var doc struct {
		Tables []struct {
			Name   string `json:"name"`
			Engine string `json:"engine"`
			Fields []struct {
				Name        string          `json:"name"`
				Kind        string          `json:"kind"`
				BitDepth    uint32          `json:"bit_depth"`
				Scale       *int64          `json:"scale"`
				Granularity []string        `json:"granularity"`
				Raw         json.RawMessage `json:"-"`
			} `json:"fields"`
		} `json:"tables"`
	}
	if err := json.Unmarshal(body, &doc); err != nil {
		return nil, &ProtocolError{What: "the schema is not readable: " + err.Error()}
	}
	out := &Schema{Raw: body}
	for _, t := range doc.Tables {
		ti := TableInfo{Name: t.Name, Engine: t.Engine}
		for _, f := range t.Fields {
			fi := FieldInfo{
				Name:        f.Name,
				Kind:        f.Kind,
				BitDepth:    f.BitDepth,
				Granularity: f.Granularity,
			}
			if f.Scale != nil {
				fi.Scale, fi.HasScale = *f.Scale, true
			}
			ti.Fields = append(ti.Fields, fi)
		}
		out.Tables = append(out.Tables, ti)
	}
	return out, nil
}

// decodeReady reads GET /ready.
func decodeReady(body []byte) (*Ready, error) {
	var doc struct {
		Status  string   `json:"status"`
		Tables  int      `json:"tables"`
		TxnID   uint64   `json:"txn_id"`
		Pages   uint64   `json:"pages"`
		Node    string   `json:"node"`
		Shards  string   `json:"shards"`
		Version string   `json:"version"`
		Wire    int      `json:"wire"`
		Serving *bool    `json:"serving"`
		Term    uint64   `json:"term"`
		Leader  string   `json:"leader"`
		Behind  []string `json:"behind"`
	}
	if err := json.Unmarshal(body, &doc); err != nil {
		return nil, &ProtocolError{What: "the readiness answer is not readable: " + err.Error()}
	}
	r := &Ready{
		Status:  doc.Status,
		Tables:  doc.Tables,
		TxnID:   doc.TxnID,
		Pages:   doc.Pages,
		Node:    doc.Node,
		Shards:  doc.Shards,
		Version: doc.Version,
		Wire:    doc.Wire,
		Term:    doc.Term,
		Leader:  doc.Leader,
		Behind:  doc.Behind,
		Raw:     body,
	}
	if doc.Serving != nil {
		r.Serving, r.HasServing = *doc.Serving, true
	}
	return r, nil
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

File: clients/python/bigdb/decode.py
# Copyright 2026 Bany
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Bytes to typed answers, mirroring `crates/big-http/src/json.rs`.

Pure: every function here takes a `RawResponse` and returns a result or raises. No sockets, no
clock, no retries - which is what lets one set of tests cover both transports.

A non-2xx never reaches a decoder. `check` turns it into the exception for its status and code
first, so a decoder only ever sees a body the server meant as an answer.
"""

from __future__ import annotations

import json
from collections.abc import Mapping
from typing import Any

from .errors import ProtocolError, classify
from .results import (
    Count,
    Extreme,
    FieldInfo,
    Group,
    Groups,
    PqlAnswer,
    ProjectionRow,
    ProjectionRows,
    Ready,
    RecordPage,
    Schema,
    SqlResult,
    Sum,
    TableInfo,
    TextResult,
    TupleGroup,
    Tuples,
    WriteResult,
)
from .transport.base import RawResponse

__all__ = [
    "check",
    "created_database",
    "created_field",
    "created_table",
    "dropped",
    "health",
    "pql",
    "ready",
    "records",
    "schema",
    "sql",
    "wrote",
]

_JSON = "application/json"


def check(raw: RawResponse) -> RawResponse:
    """The response itself, or the exception the server's refusal means."""
    if raw.ok:
        return raw
    raise classify(
        raw.status,
        raw.body,
        request_id=raw.request_id,
        retry_after=raw.retry_after,
        challenge=raw.header("www-authenticate"),
    )


def body(raw: RawResponse) -> Mapping[str, Any]:
    """A 2xx body as an object, or a `ProtocolError` naming what arrived instead.

    A 2xx that is not JSON is this client's problem to report clearly: it means a proxy answered
    or the build is not the one it thinks it is talking to, and either is worth a sentence.
    """
    check(raw)
    try:
        decoded = json.loads(raw.body)
    except (ValueError, UnicodeDecodeError) as e:
        preview = raw.body[:120].decode("utf-8", "replace")
        raise ProtocolError(
            f"a {raw.status} answer that is not JSON (content-type "
            f"{raw.content_type or 'absent'}): {preview!r}"
        ) from e
    if not isinstance(decoded, dict):
        raise ProtocolError(
            f"a {raw.status} answer that is not an object: {type(decoded).__name__}"
        )
    return decoded


# --------------------------------------------------------------------------------------------
# Probes.
# --------------------------------------------------------------------------------------------


def health(raw: RawResponse) -> bool:
    """`GET /health`, which is a constant and never authenticated."""
    return raw.ok


def ready(raw: RawResponse) -> Ready:
    """`GET /ready`. Every field optional, because a single node sends fewer than a cluster."""
    decoded = body(raw)
    behind = decoded.get("behind")
    return Ready(
        status=_str(decoded.get("status")),
        tables=_int(decoded.get("tables")),
        txn_id=_int(decoded.get("txn_id")),
        pages=_int(decoded.get("pages")),
        node=_str(decoded.get("node")),
        shards=_str(decoded.get("shards")),
        version=_str(decoded.get("version")),
        wire=_int(decoded.get("wire")),
        serving=decoded.get("serving") if isinstance(decoded.get("serving"), bool) else None,
        term=decoded.get("term") if isinstance(decoded.get("term"), int) else None,
        leader=decoded.get("leader") if isinstance(decoded.get("leader"), str) else None,
        behind=tuple(str(b) for b in behind) if isinstance(behind, list) else (),
        raw=decoded,
    )


# --------------------------------------------------------------------------------------------
# Schema.
# --------------------------------------------------------------------------------------------


def schema(raw: RawResponse) -> Schema:
    """`GET /schema`."""
    decoded = body(raw)
    tables = decoded.get("tables")
    if not isinstance(tables, list):
        raise ProtocolError("a schema with no `tables` list in it")
    return Schema(tables=tuple(_table(t) for t in tables), raw=decoded)


def _table(item: Any) -> TableInfo:
    if not isinstance(item, dict):
        raise ProtocolError(f"a table that is not an object: {type(item).__name__}")
    fields = item.get("fields")
    return TableInfo(
        name=_str(item.get("name")),
        engine=_str(item.get("engine")),
        fields=tuple(_field(f) for f in fields) if isinstance(fields, list) else (),
        raw=item,
    )


def _field(item: Any) -> FieldInfo:
    if not isinstance(item, dict):
        raise ProtocolError(f"a field that is not an object: {type(item).__name__}")
    granularity = item.get("granularity")
    scale = item.get("scale")
    return FieldInfo(
        name=_str(item.get("name")),
        # Verbatim. `signedint` here is `signed` on the way in, and normalising the two would
        # invent a third vocabulary neither side speaks.
        kind=_str(item.get("kind")),
        bit_depth=_int(item.get("bit_depth")),
        scale=scale if isinstance(scale, int) and not isinstance(scale, bool) else None,
        granularity=tuple(str(g) for g in granularity) if isinstance(granularity, list) else (),
        raw=item,
    )


# --------------------------------------------------------------------------------------------
# `/sql`.
# --------------------------------------------------------------------------------------------


def sql(raw: RawResponse) -> SqlResult | TextResult:
    """`POST /sql`, in whichever of the two shapes the statement's `FORMAT` asked for.

    **Decided by the response's content type, never by reading the statement.** A client that
    scanned SQL for a `FORMAT` clause would be a second parser to keep in step with `big_sql`,
    which is the thing every client in this repository refuses to be.
    """
    check(raw)
    if raw.content_type and raw.content_type != _JSON:
        return TextResult(
            content_type=raw.content_type,
            text=raw.body.decode("utf-8", "replace"),
        )
    decoded = body(raw)
    columns = decoded.get("columns")
    rows = decoded.get("rows")
    if not isinstance(columns, list) or not isinstance(rows, list):
        raise ProtocolError("a result set with no `columns` and `rows` in it")
    return SqlResult(
        columns=tuple(str(c) for c in columns),
        rows=tuple(tuple(row) if isinstance(row, list) else (row,) for row in rows),
        raw=decoded,
    )


# --------------------------------------------------------------------------------------------
# `/table/{t}/query` and `/table/{t}/records`.
# --------------------------------------------------------------------------------------------


def records(raw: RawResponse) -> RecordPage:
    """`GET /table/{t}/records`, which only ever answers in the record shape."""
    decoded = body(raw)
    return _record_page(decoded)


def pql(raw: RawResponse) -> PqlAnswer:
    """`POST /table/{t}/query`.

    Seven shapes and no discriminant on the wire: they are told apart by which key is present.
    Checked in a fixed order, and a shape none of them matches raises rather than handing back
    an untyped dict - a build that has not heard of an answer should say so.
    """
    return _answer(body(raw))


def _answer(decoded: Mapping[str, Any]) -> PqlAnswer:
    if "records" in decoded:
        return _record_page(decoded)
    if "count" in decoded:
        return Count(value=_int(decoded["count"]), raw=decoded)
    if "sum" in decoded:
        return Sum(value=_number(decoded["sum"]), raw=decoded)
    if "value" in decoded:
        value = decoded["value"]
        return Extreme(value=None if value is None else _number(value), raw=decoded)
    if "tuples" in decoded:
        return Tuples(tuples=tuple(_tuple(t) for t in _list(decoded["tuples"])), raw=decoded)
    if "rows" in decoded:
        return ProjectionRows(
            rows=tuple(_projection(r) for r in _list(decoded["rows"])), raw=decoded
        )
    if "groups" in decoded:
        return Groups(groups=tuple(_group(g) for g in _list(decoded["groups"])), raw=decoded)
    raise ProtocolError(
        f"this answer is in no shape this client knows: keys {sorted(decoded)!r}. "
        f"The raw body is available on the exception's cause if the server is newer than "
        f"this client."
    )


def _record_page(decoded: Mapping[str, Any]) -> RecordPage:
    ids = decoded.get("records")
    if not isinstance(ids, list):
        raise ProtocolError("a record page with no `records` list in it")
    nxt = decoded.get("next")
    return RecordPage(
        records=tuple(_int(r) for r in ids),
        next=nxt if isinstance(nxt, int) and not isinstance(nxt, bool) else None,
        raw=decoded,
    )


def _nested(value: Any) -> PqlAnswer:
    """The inner answer of a tuple or a group.

    `json::group` and the tuple writer both call `value()` recursively, so this is another
    complete answer - and a missing or non-object one is a shape this build does not know,
    which `_answer` will say so about rather than silently reading as empty.
    """
    return _answer(value if isinstance(value, dict) else {})


def _tuple(item: Any) -> TupleGroup:
    if not isinstance(item, dict):
        raise ProtocolError(f"a tuple that is not an object: {type(item).__name__}")
    keys = item.get("keys")
    return TupleGroup(
        keys=tuple(k if k is None else str(k) for k in _list(keys)),
        value=_nested(item.get("value")),
        raw=item,
    )


def _projection(item: Any) -> ProjectionRow:
    if not isinstance(item, dict):
        raise ProtocolError(f"a row that is not an object: {type(item).__name__}")
    values = item.get("values")
    return ProjectionRow(
        record=_int(item.get("record")),
        values=tuple(values) if isinstance(values, list) else (),
        raw=item,
    )


def _group(item: Any) -> Group:
    if not isinstance(item, dict):
        raise ProtocolError(f"a group that is not an object: {type(item).__name__}")
    key = item.get("key")
    return Group(
        key=None if key is None else str(key),
        row=_int(item.get("row")),
        value=_nested(item.get("value")),
        raw=item,
    )


# --------------------------------------------------------------------------------------------
# Writes and DDL.
# --------------------------------------------------------------------------------------------


def wrote(name: str) -> Any:
    """A decoder for `{"imported": n}` or `{"deleted": n}`, with its optional `missed`."""

    def decode(raw: RawResponse) -> WriteResult:
        decoded = body(raw)
        if name not in decoded:
            raise ProtocolError(f"a write answer with no `{name}` count in it")
        missed = decoded.get("missed")
        return WriteResult(
            written=_int(decoded[name]),
            # Node descriptions, not record ids: an unreachable replica does not fail a write,
            # it is named here and `POST /repair` catches it up.
            missed=tuple(str(m) for m in missed) if isinstance(missed, list) else (),
            raw=decoded,
        )

    return decode


def created_table(raw: RawResponse) -> int:
    """`{"table": id}`."""
    return _int(body(raw).get("table"))


def created_field(raw: RawResponse) -> int:
    """`{"field": id}`."""
    return _int(body(raw).get("field"))


def created_database(raw: RawResponse) -> bool:
    """`{"database": name, "created": bool}` - `False` when it was already there."""
    return bool(body(raw).get("created"))


def dropped(raw: RawResponse) -> str:
    """`{"dropped": name}`."""
    return _str(body(raw).get("dropped"))


# --------------------------------------------------------------------------------------------
# Small readers that say what was wrong rather than raising a TypeError three frames later.
# --------------------------------------------------------------------------------------------


def _int(value: Any) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ProtocolError(f"expected a whole number here, got {value!r}")
    return value


def _number(value: Any) -> int | float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ProtocolError(f"expected a number here, got {value!r}")
    return value


def _str(value: Any) -> str:
    if not isinstance(value, str):
        raise ProtocolError(f"expected a string here, got {value!r}")
    return value


def _list(value: Any) -> list[Any]:
    if not isinstance(value, list):
        raise ProtocolError(f"expected a list here, got {value!r}")
    return value
